sdk/python/feast/infra/utils/aws_utils.py

```diff
@@ -0,0 +1,286 @@
+except ImportError as e:
+    from feast.errors import FeastExtrasDependencyImportError
+
+    raise FeastExtrasDependencyImportError("aws", str(e))
+
+
+def get_redshift_data_client(aws_region: str):
+    """
+    Get the Redshift Data API Service client for the given AWS region.
+    """
+    return boto3.client("redshift-data", config=Config(region_name=aws_region))
+
+
+def get_s3_resource(aws_region: str):
+    """
+    Get the S3 resource for the given AWS region.
+    """
+    return boto3.resource("s3", config=Config(region_name=aws_region))
+
+
+def get_bucket_and_key(s3_path: str) -> Tuple[str, str]:
+    """
+    Get the S3 bucket and key given the full path.
+
+    For example get_bucket_and_key("s3://foo/bar/test.file") returns ("foo", "bar/test.file")
+
+    If the s3_path doesn't start with "s3://", it throws ValueError.
+    """
+    assert s3_path.startswith("s3://")
+    s3_path = s3_path.replace("s3://", "")
+    bucket, key = s3_path.split("/", 1)
+    return bucket, key
+
+
+def execute_redshift_statement_async(
+    redshift_data_client, cluster_id: str, database: str, user: str, query: str
+) -> dict:
+    """
+    Execute Redshift statement asynchronously. Does not wait for the query to finish.
+    Raises RedshiftCredentialsError if the statement couldn't be executed due to the validation error.
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param cluster_id: Redshift Cluster Identifier
+    :param database: Redshift Database Name
+    :param user: Redshift username
+    :param query: The SQL query to execute
+    :return: JSON response
+    """
+    try:
+        return redshift_data_client.execute_statement(
+            ClusterIdentifier=cluster_id, Database=database, DbUser=user, Sql=query,
+        )
+    except ClientError as e:
+        if e.response["Error"]["Code"] == "ValidationException":
+            raise RedshiftCredentialsError() from e
+        raise
+
+
+class RedshiftStatementNotFinishedError(Exception):
+    pass
+
+
+@retry(
+    wait=wait_exponential(multiplier=0.1, max=30),
+    retry=retry_if_exception_type(RedshiftStatementNotFinishedError),
+)
+def wait_for_redshift_statement(redshift_data_client, statement: dict) -> None:
+    """
+    Waits for the Redshift statement to finish. Raises RedshiftQueryError if the statement didn't succeed.
+
+    We use exponential backoff for checking the query state until it's not running. The backoff starts with
+    0.1 seconds and doubles exponentially until reaching 30 seconds, at which point the backoff is fixed.
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param statement: The redshift statement to wait for (result of execute_redshift_statement)
+    """
+    desc = redshift_data_client.describe_statement(Id=statement["Id"])
+    if desc["Status"] in ("SUBMITTED", "STARTED", "PICKED"):
+        raise RedshiftStatementNotFinishedError  # Retry
+    if desc["Status"] != "FINISHED":
```

Comment:
How are we handling `ABORTED` and `FAILED` states?

Reply:
They will raise RedshiftQueryError exception as documented in the docstring above.